### PR TITLE
Add ProjectDiv for L2 IntegratedGLL basis

### PR DESCRIPTION
**Pull request by pazner**
This PR adds `L2_QuadrilateralElement::ProjectDiv` and `L2_HexahedronElement::ProjectDiv` for `BasisType::IntegratedGLL`. This fixes the construction of the `DivergenceInterpolator` matrix from H(div) to L2 spaces when using this basis type.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2969](https://github.com/mfem/mfem/pull/2969) | @pazner | @tzanio | @vladotomov + @dylan-copeland | 4/15/22 | 5/13/22 | 5/15/22 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by dylan-copeland** [comment]
@pazner is there a unit test or example for testing this PR?

**Comment by pazner** [comment]
> @pazner is there a unit test or example for testing this PR?

You can use ex24 to check that this works properly. Supporting ex24 required implementing `Project(Coefficient &…` for the L2 elements with integrated basis. It also revealed a tricky bug that is a result of the caching and sharing of `Poly_1D::Basis` objects between different finite elements.

In any case, with this diff, you can test this PR with ex24:
```diff
diff --git a/examples/ex24.cpp b/examples/ex24.cpp
index 5c76ea9bd..11a73bb59 100644
--- a/examples/ex24.cpp
+++ b/examples/ex24.cpp
@@ -131,8 +131,10 @@ int main(int argc, char *argv[])
    }
    else
    {
-      trial_fec = new RT_FECollection(order-1, dim);
-      test_fec = new L2_FECollection(order-1, dim);
+      int b1 = BasisType::GaussLobatto, b2 = BasisType::IntegratedGLL;
+      int mt = FiniteElement::VALUE;
+      trial_fec = new RT_FECollection(order-1, dim, b1, b2);
+      test_fec = new L2_FECollection(order-1, dim, b2, mt);
    }
 
    FiniteElementSpace trial_fes(mesh, trial_fec);
```

Note that if you change the map type `mt` to `FiniteElement::INTEGRAL`, the solution to the variational problem will not be correct. This is because `VectorFEDivergenceIntegrator` does not support test spaces with map type `INTEGRAL`.

<details>
<summary>This patch fixes that issue (cc: @tzanio, would we want to merge this change?)</summary>

```diff
diff --git a/fem/bilininteg.cpp b/fem/bilininteg.cpp
index 93806cd1f..2c6b3ee55 100644
--- a/fem/bilininteg.cpp
+++ b/fem/bilininteg.cpp
@@ -1672,7 +1672,8 @@ void VectorFEDivergenceIntegrator::AssembleElementMatrix2(
    {
       const IntegrationPoint &ip = ir->IntPoint(i);
       trial_fe.CalcDivShape(ip, divshape);
-      test_fe.CalcShape(ip, shape);
+      Trans.SetIntPoint(&ip);
+      test_fe.CalcPhysShape(Trans, shape);
       double w = ip.weight;
       if (Q)
       {
diff --git a/fem/bilininteg.hpp b/fem/bilininteg.hpp
index 4c909a01d..0f89b590e 100644
--- a/fem/bilininteg.hpp
+++ b/fem/bilininteg.hpp
@@ -2391,8 +2391,9 @@ public:
     scalar function given by FiniteElement through standard transformation.
     Here, u is the trial function and p is the test function.
 
-    Note: the element matrix returned by AssembleElementMatrix2 does NOT depend
-    on the ElementTransformation Trans. */
+    Note: if the test space does not have map type INTEGRAL, then the element
+    matrix returned by AssembleElementMatrix2 will not depend on the
+    ElementTransformation Trans. */
 class VectorFEDivergenceIntegrator : public BilinearFormIntegrator
 {
 protected:
```
</details>

**Comment by tzanio** [comment]
> Note that if you change the map type mt to FiniteElement::INTEGRAL, the solution to the variational problem will not be correct. This is because VectorFEDivergenceIntegrator does not support test spaces with map type INTEGRAL.
> 
> This patch fixes that issue (cc: @tzanio would we want to merge this change?)

Yes please

File: fem/fe/fe_l2.cpp [hunk 1]
```diff
@@ -182,6 +187,113 @@ void L2_QuadrilateralElement::ProjectDelta(int vertex, Vector &dofs) const
...
+   if (basis1d.IsIntegratedType())
+   {
+      const IntegrationRule &ir = IntRules.Get(geom_type, order);
+      const double *gll_pts = poly1d.GetPoints(order+1, BasisType::GaussLobatto);
+
+      dofs = 0.0;
+      // Loop over subcells
+      for (int iy = 0; iy < order+1; ++iy)
+      {
+         double hy = gll_pts[iy+1] - gll_pts[iy];
```
Comment (dylan-copeland) [inline fem/fe/fe_l2.cpp:268]: const double

File: fem/fe/fe_l2.cpp [hunk 2]
```diff
@@ -182,6 +187,113 @@ void L2_QuadrilateralElement::ProjectDelta(int vertex, Vector &dofs) const
...
+            const int i = ix + iy*(order+1);
+            double hx = gll_pts[ix+1] - gll_pts[ix];
+            // Loop over subcell quadrature points
+            for (int iq = 0; iq < ir.Size(); ++iq)
+            {
+               IntegrationPoint ip = ir[iq];
+               ip.x = gll_pts[ix] + hx*ip.x;
+               ip.y = gll_pts[iy] + hy*ip.y;
+               Trans.SetIntPoint(&ip);
+               double val = coeff.Eval(Trans, ip);
```
Comment (dylan-copeland) [inline fem/fe/fe_l2.cpp:280]: const double

Comment (pazner) [reply fem/fe/fe_l2.cpp]: Done

File: fem/fe/fe_l2.cpp (hunk 1 above)
Comment (pazner) [reply fem/fe/fe_l2.cpp]: Done

**Review by dylan-copeland — APPROVED** [review]

**Comment by tzanio** [comment]
@vladotomov, can you please take a look when you get a chance?

File: fem/fe/fe_l2.cpp [hunk 3]
```diff
@@ -104,6 +107,7 @@ void L2_QuadrilateralElement::CalcShape(const IntegrationPoint &ip,
    Vector shape_x(p+1), shape_y(p+1);
 #endif
 
+   basis1d.ScaleIntegrated(map_type == VALUE);
```
Comment (vladotomov) [inline fem/fe/fe_l2.cpp:110]: Shouldn't these calls be somewhere in the constructors? Doesn't seem related to the `CalcShape` arguments.

Also, since it's relevant only for `Integrated` bases, I would put it inside `if (basis is integrated)` for better intuition.

**Comment by vladotomov** [comment]
These functions don't seem trivial, and there is no automatic check of errors / regression. Should we add something? Seems adding an option + sample run in `ex24` won't take much effort?

File: fem/fe/fe_l2.cpp (hunk 3 above)
Comment (pazner) [reply fem/fe/fe_l2.cpp:110]: The issue is that the `basis1d` objects are cached and shared between different finite element objects, so if it's set in the constructor of one finite element, it will actually have an effect on the evaluation of all other finite elements sharing the same basis type and degree.

**Comment by pazner** [comment]
> These functions don't seem trivial, and there is no automatic check of errors / regression. Should we add something? Seems adding an option + sample run in `ex24` won't take much effort?

It would be a little bit strange to add this as an option to `ex24` — I didn't implement `ProjectGrad` or `ProjectCurl` for ND and RT spaces, so only the `-p 2` flag would work properly with the integrated basis.

File: fem/fe/fe_l2.cpp (hunk 3 above)
Comment (vladotomov) [reply fem/fe/fe_l2.cpp:110]: Got it, but I still recommend using an `if` to improve clarity. It might be good also to comment somewhere that the `Integrated` basis is relevant only for `L2` FE (I assume that's the case since the calls are not made in the `H1_` FE classes).

**Review by vladotomov — APPROVED** [review]
Apologies for the delay.

**Comment by tzanio** [comment]
Merged in `next` for testing...
